File: go/muse/algorithms/sorting/shell_sort/shell_sort.go
package shell_sort

func Sort(arr []int) {
	gaps := []int{701, 301, 132, 57, 23, 10, 4, 1}
	for _, gap := range gaps {
		for i := gap; i < len(arr); i++ {
			target := arr[i]
			cursor := i
			for ; cursor >= gap; cursor -= gap {
				if arr[cursor-gap] <= target {
					break
				}
				arr[cursor] = arr[cursor-gap]
			}
			arr[cursor] = target
		}
	}
}

File: rust/muse/src/algorithms/sorting/shell_sort.rs
pub fn sort(arr: &mut [i32]) {
    let gaps = [701, 301, 132, 57, 23, 10, 4, 1];
    for gap in gaps.into_iter() {
        for i in gap..arr.len() {
            let target = arr[i];
            let mut cursor = i;
            while cursor >= gap {
                if arr[cursor - gap] <= target {
                    break;
                }
                arr[cursor] = arr[cursor - gap];
                cursor -= gap;
            }
            arr[cursor] = target;
        }
    }
}

#[cfg(test)]
mod tests;
